### ISSUE TITLE
Computing Rouge score during training

### ISSUE THREAD

**Issue by sjelassi** [PiotrNawrot/nanoT5#3]
Hi Piotr, 

Thank you for your amazing work on the nanoT5 repository. 

I am a beginner in NLP so I have been trying to learn how to run T5 thanks to your repo. I have a question regarding the computation of the rouge score. I am trying to compute it during the training of my T5 to monitor the performance of my model. 

When I set up return_attention_mask=True in your code (https://github.com/PiotrNawrot/nanoT5/blob/main/nanoT5/utils/copied_utils.py#L361), I have a mismatch of sizes in "position_bias = position_bias + mask" with position_bias being of size 512 (the input length) and the mask of size 568 (which is the before_mask_input_length )

**Question**: Do you have an advice to be able to compute the rouge score every X steps during training using your repo? 

Thanks a lot for your help !
Samy

**Comment by sjelassi** [comment]
Response from Piotr: 

Hi Samy,

Glad that you like my repo 🙂!

In my repo I assummed the following pipeline: 

First we pre-train the T5 model, possibly saving checkpoints every X steps, then we load another config (configs/task/ft.yaml) responsible for fine-tuning and therefore calculating RougeL score on SNI dataset with a specified checkpoint (model.checkpoint_path)

As you can see, I calculate the RougeL score only if the args.mode == 'ft' (https://github.com/PiotrNawrot/nanoT5/blob/main/nanoT5/utils/train_utils.py#L27).

Attention mask is needed to calculate RougeL and perform fine-tuning on SNI dataset, because the examples in the batch could have different length. During pre-training we don't need attention mask, because all examples have the same length (You can read about the specifics of the pre-training objective from the original T5 paper).

Tokenize function you linked is only used for processing the C4 dataset we use for pre-training, that's why I specify return_attention_mask=False there. The mismatch you observe is another consequence of data preprocessing for T5 pre-training objective. Shortly, what we do is: we tokenize random articles from the web of length and extract sequences of length 568. Then we mask random contiguous segments (possibly longer than 1 token), and replace them with a single sentinel (extra) mask token. That's why the final input length is 512, instead of 568. By doing this preprocessing masking we effectively reduce the number of tokens by replacing randomly picked contiguous segments with sentinel tokens.

Finally, my advice to evaluate RougeL during pre-training every X steps.

There are two steps: pre-training (C4 dataset) and fine-tuning (SNI dataset). To get a good RougeL, no matter what, you should expose your model to some examples from SNI dataset through fine-tuning. I see two options:

Run pre-training and save checkpoints every X steps by changing this argument (https://github.com/PiotrNawrot/nanoT5/blob/main/nanoT5/configs/default.yaml#L43). Then, on each checkpoint you saved run fine-tuning (I describe how to do it in the README, "Run fine-tuning" section) to get RougeL.
Based on the main.py file you could implement loading optimizers/data_loaders etc. for both pre-training and fine-tuning. Then you could call fine-tuning function every X steps while you do pre-training. 
Hope that helps! Hit me up if you have other questions.

All the best,
Piotr

**Comment by PiotrNawrot** [comment]
Thanks for adding it Samy!

Let's leave it here for now in case someone has similar question
